cachi2/core/package_managers/npm.py

```diff
@@ -476,6 +502,104 @@ def _get_npm_dependencies(
     return download_paths
 
 
+def get_type_of_dependency(dependency_version: str) -> Literal["registry", "git", "file", "https"]:
+    """Return type of dependency.
+
+    :param dependency_version: Dependency version loaded from package(-lock).json dependencies
+    """
+    url = urlparse(dependency_version)  # TODO: Do we need to support also http: ? I believe not
+    if url.scheme == "file":
+        return "file"
+    elif (
+        url.scheme == "git"
+        or url.scheme.startswith("git+")
+        or (url.scheme == "https" and url.path.endswith(".git"))
+    ):
+        return "git"
+    elif url.scheme == "https" and url.path.endswith(".tar.gz") or url.path.endswith(".tgz"):
+        return "https"
+    elif url.path.count("/") == 1:
+        # GitHub URLs are also supported in form of namespace/project e.g. "expressjs/express"
+        return "git"
+    else:
+        # Anything else should be considered as registry version or range of registry versions or latest
+        return "registry"
```

Comment:
Do we need to classify package.json version strings? I think we just need to check

1. is it from the registry
2. is it a file: dependency

I think that checking if a version is from the registry could be as simple as `url.scheme == ""`

Looking at https://github.com/npm/node-semver#advanced-range-syntax, none of the valid version specifiers contain `:` so it shouldn't be possible to confuse a semver range for a url (something that has a scheme)

And semver ranges also can't contain `/`, so extending this to handle the `namespace/project` corner case should still be easy
```python
is_registry = url.scheme == "" and "/" not in dependency_version
```

Reply:
Updated. Please check.